File: frontend/src/api/sync.ts
import { api } from './client'

export type SyncTargetType = 'obsidian' | 'notion'

export interface SyncTarget {
  id: string
  type: SyncTargetType
  name: string
  vault_path: string
  base_folder: string
  config_json: string
  enabled: boolean
  auto_sync: boolean
  is_default?: boolean
  created_at: number
  updated_at: number
}

export interface SaveSyncTargetInput {
  id?: string
  type?: SyncTargetType
  name: string
  vault_path: string
  base_folder: string
  config_json?: string
  enabled: boolean
  auto_sync: boolean
  is_default?: boolean
}

export type SyncStateStatus =
  | 'synced'
  | 'pending'
  | 'failed'
  | 'external_deleted'

export interface SyncState {
  note_id: string
  target_id: string
  external_path: string
  external_id: string
  external_url: string
  content_hash: string
  external_hash: string
  external_mtime: number | null
  last_direction: string
  last_synced_at: number | null
  status: SyncStateStatus
  error_message: string | null
}

export interface NoteSyncBinding {
  note_id: string
  target_id: string
  created_at: number
  updated_at: number
}

export interface NoteSyncBindingCandidate {
  target: SyncTarget
  state?: SyncState
}

export interface NoteSyncBindingResponse {
  binding: NoteSyncBinding | null
  target?: SyncTarget
  state?: SyncState
  candidates?: NoteSyncBindingCandidate[]
  binding_mismatch?: boolean
  default_target_missing?: boolean
  binding_required?: boolean
  bound_target_id?: string
  bound_target_name?: string
}

export interface SaveNoteSyncBindingRequest {
  target_id: string
  expected_target_id?: string
  confirm_changed_target?: boolean
}

export interface SaveNoteSyncBindingResponse {
  binding: NoteSyncBinding
  target: SyncTarget
  changed_target: boolean
}

export interface DeleteNoteSyncBindingRequest {
  expected_target_id: string
  expected_updated_at: number
}

export interface SyncResultItem {
  note_id: string
  status: string
  external_path?: string
  external_id?: string
  external_url?: string
  error_message?: string
}

export interface ObsidianBidirectionalResult {
  pushed: number
  pulled: number
  imported: number
  external_deleted: number
  failed: number
  items: SyncResultItem[]
}

export interface NotionBidirectionalResult {
  pushed: number
  pulled: number
  conflict_pulled: number
  imported: number
  external_deleted: number
  unsupported: number
  failed: number
  items: SyncResultItem[]
}

export interface ExternalDeletedNote {
  note_id: string
  title: string
  external_path: string
  last_synced_at: number | null
}

export interface LocalDirectoryEntry {
  name: string
  path: string
  modified_at: number
}

export interface LocalDirectoryList {
  current_path: string
  parent_path?: string
  entries: LocalDirectoryEntry[]
}

export interface SyncBatchResult {
  synced: number
  failed: number
  items: SyncResultItem[]
}

export interface TargetSyncResult {
  pushed: number
  pulled: number
  imported: number
  external_deleted: number
  conflict_pulled?: number
  unsupported?: number
  failed: number
  items: SyncResultItem[]
}

export async function getSyncTargets(): Promise<SyncTarget[]> {
  const res = await api.get<{ targets: SyncTarget[] }>('/api/sync/targets')
  return res.data.targets
}

export async function listLocalDirectories(
  path?: string
): Promise<LocalDirectoryList> {
  const res = await api.get<{ directory: LocalDirectoryList }>(
    '/api/system/directories',
    path ? { path } : undefined
  )
  return res.data.directory
}

export async function saveSyncTarget(
  input: SaveSyncTargetInput
): Promise<SyncTarget> {
  const { id } = input
  const body = syncTargetPayload(input)
  const res = id
    ? await api.patch<{ target: SyncTarget }>(`/api/sync/targets/${id}`, body)
    : await api.post<{ target: SyncTarget }>('/api/sync/targets', body)
  return res.data.target
}

export async function deleteSyncTarget(targetID: string): Promise<void> {
  await api.del(`/api/sync/targets/${encodeURIComponent(targetID)}`)
}

export async function testObsidianTarget(
  input: SaveSyncTargetInput
): Promise<void> {
  const body = {
    name: input.name,
    vault_path: input.vault_path,
    base_folder: input.base_folder,
    enabled: input.enabled,
    auto_sync: input.auto_sync,
  }
  await api.post<{ ok: boolean }>('/api/sync/obsidian/test', body)
}

export async function testNotionTarget(
  input: SaveSyncTargetInput
): Promise<void> {
  const body = syncTargetPayload(input) as ReturnType<
    typeof syncTargetPayload
  > & { id?: string }
  if (input.id) {
    body.id = input.id
  }
  await api.post<{ ok: boolean }>('/api/sync/notion/test', body)
}

export async function syncObsidianNote(id: string): Promise<SyncResultItem> {
  const res = await api.post<{ item: SyncResultItem }>(
    `/api/sync/obsidian/notes/${encodeURIComponent(id)}`
  )
  return res.data.item
}

export async function getNoteSyncBinding(
  id: string
): Promise<NoteSyncBindingResponse> {
  const res = await api.get<NoteSyncBindingResponse>(
    `/api/notes/${encodeURIComponent(id)}/sync-binding`
  )
  return res.data
}

export async function putNoteSyncBinding(
  id: string,
  payload: SaveNoteSyncBindingRequest
): Promise<SaveNoteSyncBindingResponse> {
  const res = await api.put<SaveNoteSyncBindingResponse>(
    `/api/notes/${encodeURIComponent(id)}/sync-binding`,
    payload
  )
  return res.data
}

export async function deleteNoteSyncBinding(
  id: string,
  payload: DeleteNoteSyncBindingRequest
): Promise<void> {
  await api.del(`/api/notes/${encodeURIComponent(id)}/sync-binding`, payload)
}

export async function syncNote(id: string): Promise<SyncResultItem> {
  const res = await api.post<{ item: SyncResultItem }>(
    `/api/sync/notes/${encodeURIComponent(id)}`
  )
  return res.data.item
}

export async function pushTarget(targetID: string): Promise<SyncBatchResult> {
  const res = await api.post<{ result: SyncBatchResult }>(
    `/api/sync/targets/${encodeURIComponent(targetID)}/push`
  )
  return res.data.result
}

export async function pullTarget(targetID: string): Promise<TargetSyncResult> {
  const res = await api.post<{ result: TargetSyncResult }>(
    `/api/sync/targets/${encodeURIComponent(targetID)}/pull`
  )
  return res.data.result
}

export async function bidirectionalTarget(
  targetID: string
): Promise<TargetSyncResult> {
  const res = await api.post<{ result: TargetSyncResult }>(
    `/api/sync/targets/${encodeURIComponent(targetID)}/bidirectional`
  )
  return res.data.result
}

export async function syncObsidianFolder(
  folderID: string
): Promise<SyncBatchResult> {
  const res = await api.post<{ result: SyncBatchResult }>(
    `/api/sync/obsidian/folders/${encodeURIComponent(folderID)}`
  )
  return res.data.result
}

export async function syncObsidianAll(): Promise<SyncBatchResult> {
  const res = await api.post<{ result: SyncBatchResult }>(
    '/api/sync/obsidian/all'
  )
  return res.data.result
}

export async function syncObsidianPull(): Promise<ObsidianBidirectionalResult> {
  const res = await api.post<{ result: ObsidianBidirectionalResult }>(
    '/api/sync/obsidian/pull'
  )
  return res.data.result
}

export async function syncObsidianBidirectional(): Promise<ObsidianBidirectionalResult> {
  const res = await api.post<{ result: ObsidianBidirectionalResult }>(
    '/api/sync/obsidian/bidirectional'
  )
  return res.data.result
}

export async function syncNotionAll(): Promise<SyncBatchResult> {
  const res = await api.post<{ result: SyncBatchResult }>(
    '/api/sync/notion/all'
  )
  return res.data.result
}

export async function syncNotionPull(): Promise<NotionBidirectionalResult> {
  const res = await api.post<{ result: NotionBidirectionalResult }>(
    '/api/sync/notion/pull'
  )
  return res.data.result
}

export async function syncNotionBidirectional(): Promise<NotionBidirectionalResult> {
  const res = await api.post<{ result: NotionBidirectionalResult }>(
    '/api/sync/notion/bidirectional'
  )
  return res.data.result
}

export async function getObsidianDeletions(): Promise<ExternalDeletedNote[]> {
  const res = await api.get<{ items: ExternalDeletedNote[] }>(
    '/api/sync/obsidian/deletions'
  )
  return res.data.items
}

export async function getNotionDeletions(): Promise<ExternalDeletedNote[]> {
  const res = await api.get<{ items: ExternalDeletedNote[] }>(
    '/api/sync/notion/deletions'
  )
  return res.data.items
}

export async function getTargetDeletions(
  targetID: string
): Promise<ExternalDeletedNote[]> {
  const res = await api.get<{ items: ExternalDeletedNote[] }>(
    `/api/sync/targets/${encodeURIComponent(targetID)}/deletions`
  )
  return res.data.items
}

export async function confirmObsidianDeletion(noteID: string): Promise<void> {
  await api.post(
    `/api/sync/obsidian/deletions/${encodeURIComponent(noteID)}/confirm`
  )
}

export async function confirmNotionDeletion(noteID: string): Promise<void> {
  await api.post(
    `/api/sync/notion/deletions/${encodeURIComponent(noteID)}/confirm`
  )
}

export async function confirmTargetDeletion(
  targetID: string,
  noteID: string
): Promise<void> {
  await api.post(
    `/api/sync/targets/${encodeURIComponent(targetID)}/deletions/${encodeURIComponent(noteID)}/confirm`
  )
}

export async function restoreObsidianDeletion(
  noteID: string
): Promise<SyncResultItem> {
  const res = await api.post<{ item: SyncResultItem }>(
    `/api/sync/obsidian/deletions/${encodeURIComponent(noteID)}/restore`
  )
  return res.data.item
}

export async function restoreNotionDeletion(
  noteID: string
): Promise<SyncResultItem> {
  const res = await api.post<{ item: SyncResultItem }>(
    `/api/sync/notion/deletions/${encodeURIComponent(noteID)}/restore`
  )
  return res.data.item
}

export async function restoreTargetDeletion(
  targetID: string,
  noteID: string
): Promise<SyncResultItem> {
  const res = await api.post<{ item: SyncResultItem }>(
    `/api/sync/targets/${encodeURIComponent(targetID)}/deletions/${encodeURIComponent(noteID)}/restore`
  )
  return res.data.item
}

export async function getNoteSyncState(
  id: string,
  target?: SyncTargetType
): Promise<SyncState | null> {
  const res = await api.get<{ state: SyncState | null }>(
    `/api/notes/${encodeURIComponent(id)}/sync-state`,
    target ? { target } : undefined
  )
  return res.data.state
}

function syncTargetPayload(input: SaveSyncTargetInput) {
  const payload: {
    type: SyncTargetType | undefined
    name: string
    vault_path: string
    base_folder: string
    config_json: string | undefined
    enabled: boolean
    auto_sync: boolean
    is_default?: boolean
  } = {
    type: input.type,
    name: input.name,
    vault_path: input.vault_path,
    base_folder: input.base_folder,
    config_json: input.config_json,
    enabled: input.enabled,
    auto_sync: input.auto_sync,
  }
  if (input.is_default !== undefined) {
    payload.is_default = input.is_default
  }
  return payload
}

File: backend/internal/model/sync.go
package model

type SyncTarget struct {
	ID         string `json:"id"`
	Type       string `json:"type"`
	Name       string `json:"name"`
	VaultPath  string `json:"vault_path"`
	BaseFolder string `json:"base_folder"`
	ConfigJSON string `json:"config_json"`
	Enabled    bool   `json:"enabled"`
	AutoSync   bool   `json:"auto_sync"`
	IsDefault  bool   `json:"is_default"`
	CreatedAt  int64  `json:"created_at"`
	UpdatedAt  int64  `json:"updated_at"`
}

type SyncState struct {
	NoteID        string  `json:"note_id"`
	TargetID      string  `json:"target_id"`
	ExternalPath  string  `json:"external_path"`
	ExternalID    string  `json:"external_id"`
	ExternalURL   string  `json:"external_url"`
	ContentHash   string  `json:"content_hash"`
	ExternalHash  string  `json:"external_hash"`
	ExternalMTime *int64  `json:"external_mtime"`
	LastDirection string  `json:"last_direction"`
	LastSyncedAt  *int64  `json:"last_synced_at"`
	Status        string  `json:"status"`
	ErrorMessage  *string `json:"error_message"`
}

type NoteSyncBinding struct {
	NoteID    string `json:"note_id"`
	TargetID  string `json:"target_id"`
	CreatedAt int64  `json:"created_at"`
	UpdatedAt int64  `json:"updated_at"`
}

type SyncExternalClaim struct {
	ExternalKey  string `json:"external_key"`
	NoteID       string `json:"note_id"`
	TargetID     string `json:"target_id"`
	ExternalType string `json:"external_type"`
	ExternalID   string `json:"external_id"`
	ExternalPath string `json:"external_path"`
	CreatedAt    int64  `json:"created_at"`
	UpdatedAt    int64  `json:"updated_at"`
}

type NoteSyncSuppression struct {
	NoteID    string `json:"note_id"`
	TargetID  string `json:"target_id"`
	Reason    string `json:"reason"`
	CreatedAt int64  `json:"created_at"`
	UpdatedAt int64  `json:"updated_at"`
}

type SyncImportTombstone struct {
	ExternalKey  string `json:"external_key"`
	TargetID     string `json:"target_id"`
	FormerNoteID string `json:"former_note_id"`
	ExternalType string `json:"external_type"`
	ExternalID   string `json:"external_id"`
	ExternalPath string `json:"external_path"`
	Reason       string `json:"reason"`
	CreatedAt    int64  `json:"created_at"`
	UpdatedAt    int64  `json:"updated_at"`
}

type SaveSyncTargetRequest struct {
	ID         string `json:"id"`
	Type       string `json:"type"`
	Name       string `json:"name" binding:"required"`
	VaultPath  string `json:"vault_path"`
	BaseFolder string `json:"base_folder"`
	ConfigJSON string `json:"config_json"`
	Enabled    bool   `json:"enabled"`
	AutoSync   bool   `json:"auto_sync"`
	IsDefault  *bool  `json:"is_default"`
}

type SaveNoteSyncBindingRequest struct {
	TargetID             string `json:"target_id" binding:"required"`
	ExpectedTargetID     string `json:"expected_target_id"`
	ConfirmChangedTarget bool   `json:"confirm_changed_target"`
}

type DeleteNoteSyncBindingRequest struct {
	ExpectedTargetID  string `json:"expected_target_id" binding:"required"`
	ExpectedUpdatedAt int64  `json:"expected_updated_at" binding:"required"`
}

type NoteSyncBindingCandidate struct {
	Target SyncTarget `json:"target"`
	State  *SyncState `json:"state,omitempty"`
}

type NoteSyncBindingResponse struct {
	Binding    *NoteSyncBinding           `json:"binding"`
	Target     *SyncTarget                `json:"target,omitempty"`
	State      *SyncState                 `json:"state,omitempty"`
	Candidates []NoteSyncBindingCandidate `json:"candidates,omitempty"`
}

type SaveNoteSyncBindingResponse struct {
	Binding       NoteSyncBinding `json:"binding"`
	Target        SyncTarget      `json:"target"`
	ChangedTarget bool            `json:"changed_target"`
}

type TargetScopedSyncRequest struct {
	ConfirmConflicts bool `json:"confirm_conflicts"`
}

type TargetScopedDeletionRequest struct {
	ExpectedTargetID string `json:"expected_target_id,omitempty"`
}

type TargetScopedDeletionResponse struct {
	Items []ExternalDeletedNote `json:"items"`
}

type SyncCompatibilityFlags struct {
	BindingMismatch      bool   `json:"binding_mismatch"`
	DefaultTargetMissing bool   `json:"default_target_missing"`
	BindingRequired      bool   `json:"binding_required"`
	BoundTargetID        string `json:"bound_target_id,omitempty"`
	BoundTargetName      string `json:"bound_target_name,omitempty"`
}

type NoteSyncStateCompatibilityRequest struct {
	TargetType string `json:"target_type,omitempty"`
	TargetID   string `json:"target_id,omitempty"`
}

type NoteSyncStateCompatibilityResponse struct {
	State  *SyncState             `json:"state"`
	Target *SyncTarget            `json:"target"`
	Flags  SyncCompatibilityFlags `json:"flags"`
}

type SyncResultItem struct {
	NoteID       string `json:"note_id"`
	Status       string `json:"status"`
	ExternalPath string `json:"external_path,omitempty"`
	ExternalID   string `json:"external_id,omitempty"`
	ExternalURL  string `json:"external_url,omitempty"`
	ErrorMessage string `json:"error_message,omitempty"`
}

type SyncBatchResult struct {
	Synced int              `json:"synced"`
	Failed int              `json:"failed"`
	Items  []SyncResultItem `json:"items"`
}

type ObsidianBidirectionalResult struct {
	Pushed          int              `json:"pushed"`
	Pulled          int              `json:"pulled"`
	Imported        int              `json:"imported"`
	ExternalDeleted int              `json:"external_deleted"`
	Failed          int              `json:"failed"`
	Items           []SyncResultItem `json:"items"`
}

type NotionBidirectionalResult struct {
	Pushed          int              `json:"pushed"`
	Pulled          int              `json:"pulled"`
	ConflictPulled  int              `json:"conflict_pulled"`
	Imported        int              `json:"imported"`
	ExternalDeleted int              `json:"external_deleted"`
	Unsupported     int              `json:"unsupported"`
	Failed          int              `json:"failed"`
	Items           []SyncResultItem `json:"items"`
}

type ExternalDeletedNote struct {
	NoteID       string `json:"note_id"`
	Title        string `json:"title"`
	ExternalPath string `json:"external_path"`
	LastSyncedAt *int64 `json:"last_synced_at"`
}
